api/src/main/java/org/pmiops/workbench/db/dao/DataSetServiceImpl.java

```diff
@@ -373,55 +554,72 @@ private String getQualifiedColumnName(Domain currentDomain, String columnName) {
   private String getColumnName(CdrBigQuerySchemaConfig.TableConfig config, String type) {
     Optional<CdrBigQuerySchemaConfig.ColumnConfig> conceptColumn =
         config.columns.stream().filter(column -> type.equals(column.domainConcept)).findFirst();
+    // TODO: move this logic to new DataSetRequestValidator class. Goal is for the service class
```

Comment:
That's an interesting design principle – why should service classes never throw exceptions? Or am I misinterpreting the comment?

Side note: we should really work towards having a place where we can document (and disagree on, and come to consensus on) over-arching code structure principles like this.

Reply:
[Umm](https://pmi-engteam.slack.com/archives/CN4AX0K9V/p1569948603007000). _Whistles_

Yeah it wasn't a clear comment. I'm OK with them letting runtime exceptions pass through, but not having them throw doesn't feel very functional and makes reasoning about the logic harder.

So I don't feel like exceptions belong in implementation/logic classes, and I only want to see throws happening at the public API boundary. But when I do throw I want them to be checked exceptions so there are no excuses and the IDE forces you to catch and deal with them. I've never worked on anything low-level enough where I felt the need to throw a RuntimeException, since I feel those are much more exceptional.

But the comment should've said we shouldn't be doing request validation, and a 400 shouldn't happen because of complicated business logic giving the user a null result. Talked with @blrubenstein the other day on the strategy for this on the front end too. We really want a 200 with an empty map and a useful message to the end user. Currently it's doing 400s for missing cohort and concept set, but not for empty column/value set.

Reply:
removed the comment.